source/extensions/http/header_validators/envoy_default/http2_header_validator.cc

```diff
@@ -0,0 +1,392 @@
+#include "source/extensions/http/header_validators/envoy_default/http2_header_validator.h"
+
+#include "source/extensions/http/header_validators/envoy_default/character_tables.h"
+
+#include "absl/container/node_hash_map.h"
+#include "absl/container/node_hash_set.h"
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Http {
+namespace HeaderValidators {
+namespace EnvoyDefault {
+
+using ::envoy::extensions::http::header_validators::envoy_default::v3::HeaderValidatorConfig;
+using ::Envoy::Http::HeaderString;
+using ::Envoy::Http::Protocol;
+using HeaderValidatorFunction = ::Envoy::Http::HeaderValidator::HeaderEntryValidationResult (
+    Http2HeaderValidator::*)(const HeaderString&);
+
+struct Http2ResponseCodeDetailValues {
+  const std::string InvalidTE = "uhv.http2.invalid_te";
+  const std::string ConnectionHeaderSanitization = "uhv.http2.connection_header_rejected";
+};
+
+using Http2ResponseCodeDetail = ConstSingleton<Http2ResponseCodeDetailValues>;
+
+/*
+ * Header validation implementation for the Http/2 codec. This class follows guidance from
+ * several RFCS:
+ *
+ * RFC 3986 <https://datatracker.ietf.org/doc/html/rfc3986> URI Generic Syntax
+ * RFC 7230 <https://datatracker.ietf.org/doc/html/rfc7230> HTTP/1.1 Message Syntax
+ * RFC 7231 <https://datatracker.ietf.org/doc/html/rfc7231> HTTP/1.1 Semantics and Content
+ * RFC 7540 <https://datatracker.ietf.org/doc/html/rfc7540> HTTP/2
+ *
+ */
+Http2HeaderValidator::Http2HeaderValidator(const HeaderValidatorConfig& config, Protocol protocol,
+                                           StreamInfo::StreamInfo& stream_info)
+    : HeaderValidator(config, protocol, stream_info) {}
+
+::Envoy::Http::HeaderValidator::HeaderEntryValidationResult
+Http2HeaderValidator::validateRequestHeaderEntry(const HeaderString& key,
+                                                 const HeaderString& value) {
+  static const absl::node_hash_map<absl::string_view, HeaderValidatorFunction> kHeaderValidatorMap{
```

Comment:
The :protocol pseudo-header is also allowed for extended CONNECT requests: https://www.rfc-editor.org/rfc/rfc8441#section-4. This is used for WebSockets. However, extended CONNECT also requires SETTINGS-based advertisement of support (https://www.rfc-editor.org/rfc/rfc8441.html#section-3). It's worth considering if we want the validator to allow :protocol with other layers validating :protocol <--> SETTINGS, or if we want to plumb an "enable extended CONNECT" bool to the validator.

Reply:
I think the server allows the client to use extended CONNECT by sending the `SETTINGS_ENABLE_CONNECT_PROTOCOL` setting. So I think from the perspective of header validator, this should be a configuration option, that turns on extended CONNECT semantics.

We actually do not do any validation right now. I think it makes sense to add TODO here and make it a separate PR to validate extended CONNECT.

Reply:
ping?

Reply:
I'm adding a TODO item here.